### ISSUE TITLE
TurboWarp uses so much RAM when playing heavy projects

### ISSUE THREAD

**Issue by Alestore** [TurboWarp/packager#492]
While I am running my TurboWarp game (270 MB) I notice that the RAM is occupied between 4 and 6 GB. I don't know how such a thing is possible,  I read a previous post here on packager issues and it was said that RAM usage is not under anyone's control. If there is any way to fix or optimize this, write here please.

**Comment by GarboMuffin** [comment]
Has it always been this bad or is this a recent regression?

**Comment by Alestore** [comment]
Always been this bad

**Comment by GarboMuffin** [comment]
Which version of the packager are you using? Starting a few days ago the online version received some optimizations to bitmap costumes; does your project use many of those?

**Comment by Alestore** [comment]
I'm using the online version (I think is always updated so I use that version). My project use lots of SVG files so I think that's the problem.

**Comment by GarboMuffin** [comment]
Next update will fix a memory leak in the plain HTML environment if you use that

I'm also working on something else that should save memory by garbage collecting the original costume data once it's no longer needed (the packager doesn't have to worry about the editor after all)

Consider posting your project publicly here or emailing it to contact@turbowarp.org and I can take a deeper look into your specific project

**Comment by Alestore** [comment]
Ok thanks. Currently I'm using Electron at 64-Bit for Windows and eventually Linux. I'll send you an email with the project today.

**Comment by GarboMuffin** [comment]
see email

**Comment by Alestore** [comment]
I tried to modify the sounds, especially those of the Stage and the heavier ones seen in the table you sent me by doing a double reverse to turn them into mono audio. The larger files, which are longer than 3:40 or 4 minutes, did not let me edit them but for the others I did everything. While I was doing it I saw that the RAM was gradually recovered, I think I have gained at least 2GB of RAM.

**Comment by GarboMuffin** [comment]
Even better than converting sounds to mono is to remove sounds that you don't actually use, of course

**Comment by Alestore** [comment]
I was thinking, isn't there a way to make the program that comes out of the packager (in my case 64-bit Windows Electron) read the audio files directly to the disk instead of putting them in RAM? It really takes up too much space and isn't even that optimizable.
If this is not possible, couldn't I just convert the audio files (especially the heavier ones) to 128kb .mp3 files?

**Comment by GarboMuffin** [comment]
> read the audio files directly to the disk instead of putting them in RAM

I wish it were that simple

> 128kb .mp3 files?

To be clear, the problem is the decoded audio data which is probably primarily influenced by sample rate and length. If you're okay with your sounds sounding worse, you could try manually downsampling them.

I would suggest waiting before you go that far. Once I finish some sound editor updates (maybe ~Tuesday?), you'll be able to convert some of the larger stereo sounds to mono.

>  Edit: I changed the Project ID and seems like to work for the frame drops.

I assume by this you mean "I changed the project ID and the frame drops went away"? That would probably indicate that your original project ID had some unexpected cloud variable values that your project isn't handling properly.

A few comments ago I mentioned some upcoming optimizations that will save you 500MB. You can now enable this in Advanced options > "Use "packaged runtime" mode (experimental)". Eventually the option will be removed and this will be the default. If I recall you have some crazy custom JS going on in your project, so consider testing if your custom JS still works in this mode.

**Comment by Alestore** [comment]
Ok thanks I'll try it and I'll let you know. I will wait for the sound editor update.

**Comment by GarboMuffin** [comment]
https://staging.turbowarp.org/editor will let you edit sounds regardless of the size (among other sound editor updates), although, like in Scratch, they may be downsampled and sound worse

**Comment by Alestore** [comment]
thank you I'll do some backups and I'll try this evening. Do you think this will actually work for gaining some RAM? for now, with the newest packaged runtime option from the packager, the game takes about 4,5 GB of RAM in total. I'll modify every single audio file in the project this evening and I'll let you know here. Thank you for your time!

**Comment by Alestore** [comment]
Is this normal?
When I edit audio files from stereo to change them to mono they take up more space instead of taking less (as it says).

Original sound:
![image](https://user-images.githubusercontent.com/82514000/183706917-cf5de32e-9298-4690-b613-9c87423ba791.png)

When edited with double reverse:
![image](https://user-images.githubusercontent.com/82514000/183706966-839fc6b3-bd3e-47e8-a518-3d7a1b534e99.png)

**Comment by GarboMuffin** [comment]
Raw asset size will be larger but decoded size will be smaller. Yes, it is strange. Just ignore that warning and see if the project ends up using less memory in the packager.

**Comment by GarboMuffin** [comment]
Scratch's audio editor is only capable of generating uncompressed WAV files. You could use an external audio editor to convert your stereo mp3 files to mono mp3 files to keep the file size low and to have more control over the sample rate.

**Comment by Alestore** [comment]
> Raw asset size will be larger (but it should compress pretty well in sb3/zip) but decoded size will be smaller. Yes, it is strange. Just ignore that warning and see if the project ends up using less memory in the packager.

Yes it worked. The packaged file in Electron now use only 3 GB of RAM instead of 5. I gained 500 MB by the packaged compiled option and the rest by editing audio files. Thanks a lot!

**Comment by Alestore** [comment]
Regarding the HTML export I still have this problem.

![image](https://user-images.githubusercontent.com/82514000/183738009-468e1bd5-0b15-4e04-bbdb-f057e2f30b32.png)
